### PR TITLE
Add info about type hints rendering to the docstrings section

### PR DESCRIPTION
**Pull request by greschd**
Adds a note to the `docstrings.rst` about how type hints can be rendered in the parameter description instead of the function signature.

### CODE REVIEW

**Review by jorgepiloto — APPROVED** [review]
Thanks for addressing this, @greschd. Approving these changes 🚀

File: doc/source/doc-style/docstrings.rst [hunk 1]
```diff
@@ -227,6 +227,32 @@ consider the following function:
...
+Type Hints
+==========
+
+By default, Sphinx renders `type hints <https://peps.python.org/pep-0484/>`_ as part
+of the function signature. This can become difficult to read, because the signature
+becomes very long.
+
+Instead, we can render the type hints as part of each parameter's description. For
+this, the ``sphinx.ext.autodoc.typehints``, ``sphinx.ext.napoleon``, and
+``numpydoc`` extensions need to be combined as follows (in ``conf.py``):
```
Comment (PipKat) [inline doc/source/doc-style/docstrings.rst:239]:
```suggestion
Instead, we can render the type hints as part of each parameter's description. To
accomplish this, the ``sphinx.ext.autodoc.typehints``, ``sphinx.ext.napoleon``, and
``numpydoc`` extensions must be combined in the ``conf.py`` file as follows:
```

File: doc/source/doc-style/docstrings.rst [hunk 2]
```diff
@@ -227,6 +227,32 @@ consider the following function:
...
+      ...
+      "sphinx.ext.autodoc.typehints",
+      "sphinx.ext.napoleon",
+      "numpydoc",
+      ...
+  ]
+  autodoc_typehints = "description"
+
+Note that the order in which the extensions are included matters.
+
```
Comment (PipKat) [inline doc/source/doc-style/docstrings.rst:253]:
```suggestion
```

File: doc/source/doc-style/docstrings.rst [hunk 3]
```diff
@@ -102,7 +102,7 @@ Parameters
 ----------
 Functions and class methods may have parameters in their signatures. All these
```
[the remark this answers is not in the document]
Comment (PipKat) [reply doc/source/doc-style/docstrings.rst:103]: I still don't know how to distinguish between methods and functions.

File: doc/source/doc-style/docstrings.rst [hunk 4]
```diff
@@ -143,17 +143,17 @@ Here is an example of a ``Parameters`` section for a class in PyAEDT:
     student_version : bool, optional
         Whether to enable the student version of AEDT. The default
         is ``False``.
-    
+
 The name of each parameter is followed by a space, a colon, a space, and then
 the data type. A parameter is optional if its keyword argument has a default shown
 in the function signature. For an optional parameter, the data type is followed by a
```
Comment (PipKat) [inline doc/source/doc-style/docstrings.rst:149]:
```suggestion
the data type. A parameter is optional if its keyword argument displays a default
in the function signature. For an optional parameter, the data type is followed by a
```

File: doc/source/doc-style/docstrings.rst (hunk 2 above)
Comment (akaszynski) [reply doc/source/doc-style/docstrings.rst:253]: We might consider keeping this and having this within a `note` directive:

```suggestion
.. note::
   The order in which the extensions are included matters.

```

**Review by PipKat — APPROVED** [review]

File: doc/source/doc-style/docstrings.rst (hunk 2 above)
Comment (PipKat) [reply doc/source/doc-style/docstrings.rst:253]: @akaszynski Keeping it is OK with me--I just changed the above to say that had to appear in the following order so viewed the note as redundant. However, it emphasizes the fact, which may be your intent.

File: doc/source/doc-style/docstrings.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (greschd) [reply doc/source/doc-style/docstrings.rst:103]: Indeed, the two are very similar! To be precise with the language:

- Function is the general term, it includes both *methods* and *free functions*:
- *methods* are functions which are "attached" to a class
- *free functions* in contrast do not belong to a class

The sentence above also mentions "class methods" -- which is again a subset of "methods"; they are attached to a class, but can be called from the class directly, as opposed to "instance methods", which need an _instance_ of a class.

In more relaxed / sloppy speech, "function" can refer to "free function" if it's clear from context.

So again, as a tree

- Functions
  - Free functions
  - Methods
    - Class methods
    - Instance methods
    - Static methods

In general, the [Python glossary](https://docs.python.org/3/glossary.html) is a good place to look up such terms. It's a bit annoying, but each programming language can have slightly different ways in which things are named.

Long story short: For the sentence above, I think it would be sufficient to say

``Functions can have parameters...``

To be pedantic, classes _also_ have signatures, but that is the same as the signature of their constructor (a special method).

File: doc/source/doc-style/docstrings.rst (hunk 4 above)
Comment (greschd) [reply doc/source/doc-style/docstrings.rst:149]:
```suggestion
the data type. A parameter is optional if a default is defined in its signature. 
For an optional parameter, the data type is followed by a
```

In this sentence, the "keyword argument" seems misplaced to me. Not all parameters are keyword arguments, see https://docs.python.org/3/glossary.html#term-argument

Comment (greschd) [reply doc/source/doc-style/docstrings.rst:149]: @PipKat I'm not sure if we should mix general edits to this page together with the addition of the type hint section.

Note that the changes I've made to the rest of the document are just the removal of trailing whitespace.

If the changes are small enough I guess we can do it in one go, but I wouldn't want to delay this PR with other changes.

File: doc/source/doc-style/docstrings.rst (hunk 2 above)
Comment (greschd) [reply doc/source/doc-style/docstrings.rst]: Ah, seems I messed this up by prematurely accepting the edits. I'll revert it tomorrow, when I'm in front of my usual PC.

@PipKat if we keep the note, would you lean towards keeping the edited version of the sentence above, or revert it to the original?

Comment (PipKat) [reply doc/source/doc-style/docstrings.rst]: The note directive will definitely put more emphasis on this information. If calling it out is important, use the note!

File: doc/source/doc-style/docstrings.rst (hunk 4 above)
Comment (PipKat) [reply doc/source/doc-style/docstrings.rst:149]: Feel free to go back to your original text on these lines. The number of edits are small--so I'd suggest making them.

File: doc/source/doc-style/docstrings.rst (hunk 2 above)
Comment (greschd) [reply doc/source/doc-style/docstrings.rst]: Ok, yeah I think we should use the note. In that case should we revert to the original 
```
Instead, we can render the type hints as part of each parameter's description. For
this, the ``sphinx.ext.autodoc.typehints``, ``sphinx.ext.napoleon``, and
``numpydoc`` extensions need to be combined as follows (in ``conf.py``):
```
or use 
```
Instead, we can render the type hints as part of each parameter's description. To
accomplish this, the ``sphinx.ext.autodoc.typehints``, ``sphinx.ext.napoleon``, and
``numpydoc`` extensions must be combined in the ``conf.py`` file in this order:
```

To me, the latter version now looks somewhat redundant, but wanted to hear your opinion on that.

Comment (PipKat) [reply doc/source/doc-style/docstrings.rst]: I'd do the last--remove "in this order" from the suggested edit.
